File: go/provenancemark/generator.go
package provenancemark

import (
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"

	bcenvelope "github.com/nickel-blockchaincommons/bcenvelope-go"
	bcrand "github.com/nickel-blockchaincommons/bcrand-go"
	dcbor "github.com/nickel-blockchaincommons/dcbor-go"
)

// ProvenanceMarkGenerator produces sequential provenance marks.
type ProvenanceMarkGenerator struct {
	res      ProvenanceMarkResolution
	seed     ProvenanceSeed
	chainID  []byte
	nextSeq  uint32
	rngState RngState
}

// NewProvenanceMarkGenerator constructs a generator from explicit state.
func NewProvenanceMarkGenerator(
	res ProvenanceMarkResolution,
	seed ProvenanceSeed,
	chainID []byte,
	nextSeq uint32,
	rngState RngState,
) (ProvenanceMarkGenerator, error) {
	if len(chainID) != res.LinkLength() {
		return ProvenanceMarkGenerator{}, newInvalidChainIDLength(res.LinkLength(), len(chainID))
	}
	return ProvenanceMarkGenerator{
		res:      res,
		seed:     seed,
		chainID:  cloneBytes(chainID),
		nextSeq:  nextSeq,
		rngState: rngState,
	}, nil
}

// NewProvenanceMarkGeneratorWithSeed derives a generator from the given seed.
func NewProvenanceMarkGeneratorWithSeed(res ProvenanceMarkResolution, seed ProvenanceSeed) ProvenanceMarkGenerator {
	seedBytes := seed.Bytes()
	digest1 := SHA256(seedBytes[:])
	chainID := cloneBytes(digest1[:res.LinkLength()])
	digest2 := SHA256(digest1[:])
	generator, err := NewProvenanceMarkGenerator(res, seed, chainID, 0, RngStateFromBytes(digest2))
	if err != nil {
		panic(err)
	}
	return generator
}

// NewProvenanceMarkGeneratorWithPassphrase derives a generator from a passphrase.
func NewProvenanceMarkGeneratorWithPassphrase(res ProvenanceMarkResolution, passphrase string) ProvenanceMarkGenerator {
	return NewProvenanceMarkGeneratorWithSeed(res, NewProvenanceSeedWithPassphrase(passphrase))
}

// NewProvenanceMarkGeneratorUsing derives a generator from randomness.
func NewProvenanceMarkGeneratorUsing(res ProvenanceMarkResolution, rng bcrand.RandomNumberGenerator) ProvenanceMarkGenerator {
	return NewProvenanceMarkGeneratorWithSeed(res, NewProvenanceSeedUsing(rng))
}

// NewRandomProvenanceMarkGenerator creates a generator using secure randomness.
func NewRandomProvenanceMarkGenerator(res ProvenanceMarkResolution) ProvenanceMarkGenerator {
	return NewProvenanceMarkGeneratorWithSeed(res, NewProvenanceSeed())
}

// Res returns the generator resolution.
func (g ProvenanceMarkGenerator) Res() ProvenanceMarkResolution {
	return g.res
}

// Seed returns the generator seed.
func (g ProvenanceMarkGenerator) Seed() ProvenanceSeed {
	return g.seed
}

// ChainID returns the chain identifier.
func (g ProvenanceMarkGenerator) ChainID() []byte {
	return cloneBytes(g.chainID)
}

// NextSeq returns the next sequence number that will be generated.
func (g ProvenanceMarkGenerator) NextSeq() uint32 {
	return g.nextSeq
}

// RngState returns the current deterministic RNG state.
func (g ProvenanceMarkGenerator) RngState() RngState {
	return g.rngState
}

// Next emits the next mark and mutates the generator state.
func (g *ProvenanceMarkGenerator) Next(date dcbor.Date, info any) ProvenanceMark {
	if g == nil {
		panic("nil ProvenanceMarkGenerator")
	}
	rng := Xoshiro256StarStarFromData(g.rngState.Bytes())

	seq := g.nextSeq
	g.nextSeq++

	var key []byte
	if seq == 0 {
		key = cloneBytes(g.chainID)
	} else {
		key = rng.NextBytes(g.res.LinkLength())
		g.rngState = RngStateFromBytes(rng.Data())
	}

	nextRng := rng.Clone()
	nextKey := nextRng.NextBytes(g.res.LinkLength())

	mark, err := NewProvenanceMark(g.res, key, nextKey, g.chainID, seq, date, info)
	if err != nil {
		panic(err)
	}
	return mark
}

// String returns the display form.
func (g ProvenanceMarkGenerator) String() string {
	return fmt.Sprintf(
		"ProvenanceMarkGenerator(chainID: %s, res: %s, seed: %s, nextSeq: %d, rngState: %s)",
		hex.EncodeToString(g.chainID),
		g.res,
		g.seed.Hex(),
		g.nextSeq,
		g.rngState.Hex(),
	)
}

// Equal reports whether two generators have identical state.
func (g ProvenanceMarkGenerator) Equal(other ProvenanceMarkGenerator) bool {
	return g.res == other.res &&
		g.seed == other.seed &&
		compareBytes(g.chainID, other.chainID) == 0 &&
		g.nextSeq == other.nextSeq &&
		g.rngState == other.rngState
}

// ToEnvelope converts the generator to an envelope.
func (g ProvenanceMarkGenerator) ToEnvelope() *bcenvelope.Envelope {
	return bcenvelope.NewEnvelope(dcbor.ToByteString(g.chainID)).
		AddType("provenance-generator").
		AddAssertion("res", g.res.ToCBOR()).
		AddAssertion("seed", g.seed.ToCBOR()).
		AddAssertion("next-seq", uint64(g.nextSeq)).
		AddAssertion("rng-state", g.rngState.ToCBOR())
}

// ProvenanceMarkGeneratorFromEnvelope decodes a generator from an envelope.
func ProvenanceMarkGeneratorFromEnvelope(envelope *bcenvelope.Envelope) (ProvenanceMarkGenerator, error) {
	if envelope == nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(errors.New("nil envelope"))
	}
	if err := envelope.CheckType("provenance-generator"); err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	chainID, err := envelope.Subject().TryByteString()
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	const expectedKeyCount = 5
	if len(envelope.Assertions()) != expectedKeyCount {
		return ProvenanceMarkGenerator{}, newExtraKeys(expectedKeyCount, len(envelope.Assertions()))
	}

	resLeaf, err := envelope.ObjectForPredicate("res")
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	resCBOR, err := resLeaf.TryLeaf()
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	res, err := DecodeProvenanceMarkResolution(resCBOR)
	if err != nil {
		return ProvenanceMarkGenerator{}, err
	}

	seedLeaf, err := envelope.ObjectForPredicate("seed")
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	seedCBOR, err := seedLeaf.TryLeaf()
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	seed, err := DecodeProvenanceSeed(seedCBOR)
	if err != nil {
		return ProvenanceMarkGenerator{}, err
	}

	nextSeqLeaf, err := envelope.ObjectForPredicate("next-seq")
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	nextSeqCBOR, err := nextSeqLeaf.TryLeaf()
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	nextSeq, err := dcbor.DecodeUInt32(nextSeqCBOR)
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapCBORError(err)
	}

	rngStateLeaf, err := envelope.ObjectForPredicate("rng-state")
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	rngStateCBOR, err := rngStateLeaf.TryLeaf()
	if err != nil {
		return ProvenanceMarkGenerator{}, wrapEnvelopeError(err)
	}
	rngState, err := DecodeRngState(rngStateCBOR)
	if err != nil {
		return ProvenanceMarkGenerator{}, err
	}

	return NewProvenanceMarkGenerator(res, seed, chainID, nextSeq, rngState)
}

type provenanceMarkGeneratorJSON struct {
	Res      ProvenanceMarkResolution `json:"res"`
	Seed     ProvenanceSeed           `json:"seed"`
	ChainID  string                   `json:"chainID"`
	NextSeq  uint32                   `json:"nextSeq"`
	RngState RngState                 `json:"rngState"`
}

// MarshalJSON encodes the generator in its public JSON form.
func (g ProvenanceMarkGenerator) MarshalJSON() ([]byte, error) {
	return json.Marshal(provenanceMarkGeneratorJSON{
		Res:      g.res,
		Seed:     g.seed,
		ChainID:  SerializeBase64(g.chainID),
		NextSeq:  g.nextSeq,
		RngState: g.rngState,
	})
}

// UnmarshalJSON decodes the generator from its public JSON form.
func (g *ProvenanceMarkGenerator) UnmarshalJSON(data []byte) error {
	var payload provenanceMarkGeneratorJSON
	if err := json.Unmarshal(data, &payload); err != nil {
		return err
	}
	chainID, err := DeserializeBase64(payload.ChainID)
	if err != nil {
		return err
	}
	decoded, err := NewProvenanceMarkGenerator(payload.Res, payload.Seed, chainID, payload.NextSeq, payload.RngState)
	if err != nil {
		return err
	}
	*g = decoded
	return nil
}

var _ json.Marshaler = ProvenanceMarkGenerator{}
var _ json.Unmarshaler = (*ProvenanceMarkGenerator)(nil)

File: python/provenance-mark/src/provenance_mark/_generator.py
"""Sequential provenance mark generator."""

from __future__ import annotations

import json

from bc_envelope import Envelope
from dcbor import Date

from ._crypto_utils import sha256
from ._error import Error
from ._mark import ProvenanceMark
from ._resolution import ProvenanceMarkResolution, resolution_from_cbor, resolution_to_cbor
from ._rng_state import RngState
from ._seed import ProvenanceSeed
from ._util import b64decode, b64encode
from ._xoshiro256starstar import Xoshiro256StarStar


class ProvenanceMarkGenerator:
    """Stateful generator for sequential provenance marks."""

    __slots__ = ("_res", "_seed", "_chain_id", "_next_seq", "_rng_state")

    def __init__(
        self,
        res: ProvenanceMarkResolution,
        seed: ProvenanceSeed,
        chain_id: bytes,
        next_seq: int,
        rng_state: RngState,
    ) -> None:
        expected = res.link_length()
        if len(chain_id) != expected:
            raise Error(
                "InvalidChainIdLength",
                f"invalid chain ID length: expected {expected}, got {len(chain_id)}",
            )
        self._res = res
        self._seed = seed
        self._chain_id = bytes(chain_id)
        self._next_seq = next_seq
        self._rng_state = rng_state

    @staticmethod
    def new(
        res: ProvenanceMarkResolution,
        seed: ProvenanceSeed,
        chain_id: bytes | bytearray,
        next_seq: int,
        rng_state: RngState,
    ) -> ProvenanceMarkGenerator:
        return ProvenanceMarkGenerator(res, seed, bytes(chain_id), next_seq, rng_state)

    @staticmethod
    def new_with_seed(
        res: ProvenanceMarkResolution,
        seed: ProvenanceSeed,
    ) -> ProvenanceMarkGenerator:
        digest1 = sha256(seed.to_bytes())
        chain_id = digest1[: res.link_length()]
        digest2 = sha256(digest1)
        return ProvenanceMarkGenerator.new(
            res,
            seed,
            chain_id,
            0,
            RngState.from_bytes(digest2),
        )

    @staticmethod
    def new_with_passphrase(
        res: ProvenanceMarkResolution,
        passphrase: str,
    ) -> ProvenanceMarkGenerator:
        return ProvenanceMarkGenerator.new_with_seed(
            res,
            ProvenanceSeed.new_with_passphrase(passphrase),
        )

    @staticmethod
    def new_using(
        res: ProvenanceMarkResolution,
        rng,
    ) -> ProvenanceMarkGenerator:
        return ProvenanceMarkGenerator.new_with_seed(res, ProvenanceSeed.new_using(rng))

    @staticmethod
    def new_random(res: ProvenanceMarkResolution) -> ProvenanceMarkGenerator:
        return ProvenanceMarkGenerator.new_with_seed(res, ProvenanceSeed.new())

    def res(self) -> ProvenanceMarkResolution:
        return self._res

    def seed(self) -> ProvenanceSeed:
        return self._seed

    def chain_id(self) -> bytes:
        return self._chain_id

    def next_seq(self) -> int:
        return self._next_seq

    def rng_state(self) -> RngState:
        return self._rng_state

    def next(self, date: Date, info: object | None = None) -> ProvenanceMark:
        rng = Xoshiro256StarStar.from_data(self._rng_state.to_bytes())
        seq = self._next_seq
        self._next_seq += 1

        if seq == 0:
            key = self._chain_id
        else:
            key = rng.next_bytes(self._res.link_length())
            self._rng_state = RngState.from_bytes(rng.to_data())

        next_rng = rng.clone()
        next_key = next_rng.next_bytes(self._res.link_length())
        return ProvenanceMark.new(
            self._res,
            key,
            next_key,
            self._chain_id,
            seq,
            date,
            info,
        )

    def to_envelope(self) -> Envelope:
        return (
            Envelope(self._chain_id)
            .add_type("provenance-generator")
            .add_assertion("res", resolution_to_cbor(self._res))
            .add_assertion("seed", self._seed.to_cbor())
            .add_assertion("next-seq", self._next_seq)
            .add_assertion("rng-state", self._rng_state.to_cbor())
        )

    @staticmethod
    def from_envelope(envelope: Envelope) -> ProvenanceMarkGenerator:
        envelope.check_type("provenance-generator")
        chain_id = envelope.subject().try_leaf().try_byte_string()
        if len(envelope.assertions()) != 5:
            raise Error(
                "ExtraKeys",
                f"wrong number of keys: expected 5, got {len(envelope.assertions())}",
            )
        return ProvenanceMarkGenerator.new(
            resolution_from_cbor(envelope.object_for_predicate("res").try_leaf()),
            ProvenanceSeed.from_cbor(envelope.object_for_predicate("seed").try_leaf()),
            chain_id,
            envelope.object_for_predicate("next-seq").try_leaf().try_int(),
            RngState.from_cbor(envelope.object_for_predicate("rng-state").try_leaf()),
        )

    def to_json(self) -> dict[str, object]:
        return {
            "res": int(self._res),
            "seed": self._seed.to_json(),
            "chainID": b64encode(self._chain_id),
            "nextSeq": self._next_seq,
            "rngState": self._rng_state.to_json(),
        }

    @staticmethod
    def from_json(value: str | dict[str, object]) -> ProvenanceMarkGenerator:
        payload = value if isinstance(value, dict) else json.loads(value)
        return ProvenanceMarkGenerator.new(
            ProvenanceMarkResolution(int(payload["res"])),
            ProvenanceSeed.from_json(str(payload["seed"])),
            b64decode(str(payload["chainID"])),
            int(payload["nextSeq"]),
            RngState.from_json(str(payload["rngState"])),
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ProvenanceMarkGenerator):
            return NotImplemented
        return (
            self._res == other._res
            and self._seed == other._seed
            and self._chain_id == other._chain_id
            and self._next_seq == other._next_seq
            and self._rng_state == other._rng_state
        )

    def __repr__(self) -> str:
        return (
            "ProvenanceMarkGenerator("
            f"chainID: {self._chain_id.hex()}, "
            f"res: {self._res}, "
            f"seed: {self._seed.hex()}, "
            f"nextSeq: {self._next_seq}, "
            f"rngState: {self._rng_state.hex()})"
        )

    __str__ = __repr__

